File: src/C++/sub_matrix.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-16 19:37:13
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-23 19:20:27
@FilePath: \asm_matrix_benchmark\src\C++\sub_matrix.cpp
@Description: sub matrix c++ code
*************************************************************/
#include "matrix_cpp.h"
#include <stdexcept>

/***********************************************************
@description: sub matrix
@return {*}
*************************************************************/
MatrixInt MatrixInt::operator-(const MatrixInt &other) const
{
    if (rows != other.rows || cols != other.cols)
    {
        throw std::invalid_argument("operator-: dimension mismatch");
    }

    MatrixInt res(rows, cols);
    for (size_t i = 0; i < rows * cols; i++)
    {
        res.data[i] = data[i] - other.data[i];
    }
    return res;
}

/***********************************************************
@description: -=
@return {*}
*************************************************************/
MatrixInt &MatrixInt::operator-=(const MatrixInt &other)
{
    if (rows != other.rows || cols != other.cols)
    {
        throw std::invalid_argument("operator-=: dimension mismatch");
    }

    for (size_t i = 0; i < rows * cols; i++)
    {
        data[i] -= other.data[i];
    }
    return *this;
}

File: src/JavaScript/sub_matrix.js
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-29 20:10:32
@LastEditors: GitHubCopilot
@LastEditTime: 2026-06-29 20:10:32
@FilePath: \asm_matrix_benchmark\src\JavaScript\sub_matrix.js
@Description: sub matrix JavaScript code
*************************************************************/

"use strict";

const { checkMatrix, sameShape, buildTypedMatrix } = require("./common");

/**
 * Generic element-wise subtraction.
 * @param {{rows:number, cols:number, data:number[]}} m1
 * @param {{rows:number, cols:number, data:number[]}} m2
 * @param {"int"|"float"|"double"} type
 * @returns {{rows:number, cols:number, data:number[]}|null}
 */
function subMatrix(m1, m2, type) {
  if (!checkMatrix(m1, "m1") || !checkMatrix(m2, "m2")) {
    return null;
  }
  if (!sameShape(m1, m2)) {
    return null;
  }

  const data = new Array(m1.rows * m1.cols);
  for (let i = 0; i < data.length; i++) {
    data[i] = m1.data[i] - m2.data[i];
  }

  return buildTypedMatrix(m1.rows, m1.cols, data, type);
}

function sub_matrix_int(m1, m2) {
  return subMatrix(m1, m2, "int");
}

function sub_matrix_float(m1, m2) {
  return subMatrix(m1, m2, "float");
}

function sub_matrix_double(m1, m2) {
  return subMatrix(m1, m2, "double");
}

module.exports = {
  sub_matrix_int,
  sub_matrix_float,
  sub_matrix_double,
};
